Graphics/GraphicsTools/src/ShaderPreprocessor.cpp

```diff
@@ -0,0 +1,198 @@
+        {
+            case None:
+            {
+                if (NextHash && NextHash < pBuffer + Offset)
+                    NextHash = static_cast<const char*>(memchr(pBuffer + Offset, '#', BufferSize - Offset));
+
+                if (NextHash == nullptr)
+                    return;
+
+                if (NextSlash && NextSlash < pBuffer + Offset)
+                    NextSlash = static_cast<const char*>(memchr(pBuffer + Offset, '/', BufferSize - Offset));
+
+                if (NextSlash && NextSlash < NextHash)
+                {
+                    Offset = NextSlash - pBuffer;
+                    if (pBuffer[Offset + 1] == '/')
+                    {
+                        Offset = static_cast<const char*>(memchr(pBuffer + Offset, '\n', BufferSize - Offset)) - pBuffer;
+                    }
+                    else if (pBuffer[Offset + 1] == '*')
+                    {
+                        do
+                        {
+                            const char* endSlash = static_cast<const char*>(memchr(pBuffer + Offset + 1, '/', BufferSize - Offset));
+                            if (!endSlash) return;
+                            Offset = endSlash - pBuffer;
+                        } while (pBuffer[Offset - 1] != '*');
+                    }
+                }
+                else
+                {
+                    Offset            = NextHash - pBuffer;
+                    PreprocessorState = AfterHash;
+                }
+            }
+            break;
+            case AfterHash:
+                if (!isspace(pBuffer[Offset]))
+                {
+                    if (strncmp(pBuffer + Offset, "include", 7) == 0)
+                    {
+                        PreprocessorState = AfterInclude;
+                        Offset += 6;
+                    }
+                    else
+                    {
+                        PreprocessorState = None;
+                    }
+                }
+                break;
+            case AfterInclude:
+
+                if (!isspace(pBuffer[Offset]))
+                {
+                    if (pBuffer[Offset] == '"')
+                    {
+                        Start             = Offset + 1;
+                        PreprocessorState = InsideIncludeBrackets;
+                    }
+                    else
+                    {
+                        PreprocessorState = None;
+                    }
+                }
+                break;
+            case InsideIncludeBrackets:
+                switch (pBuffer[Offset])
+                {
+                    case '\n':
+                        PreprocessorState = None; // Buggy code, skip over this include.
+                        break;
+                    case '\"':
+                        IncludeHandler(std::string(&pBuffer[Start], &pBuffer[Offset]));
+                        PreprocessorState = None;
+                        break;
+                }
+                break;
```

Comment:
Please add comments for parser operation - they will be very helpful to show what the parser is doing